packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,724 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'package:process/process.dart';
+
+import '../application_package.dart';
+import '../base/common.dart';
+import '../base/context.dart';
+import '../base/io.dart';
+import '../base/logger.dart';
+import '../base/process.dart';
+import '../base/utils.dart';
+import '../build_info.dart';
+import '../bundle.dart';
+import '../convert.dart';
+import '../device.dart';
+import '../features.dart';
+import '../project.dart';
+import '../protocol_discovery.dart';
+import 'custom_device_config.dart';
+import 'custom_device_workflow.dart';
+import 'custom_devices_config.dart';
+
+List<String> interpolateCommand(
+  List<String> command,
+  Map<String, String> replacementValues, {
+  Map<String, String> additionalReplacementValues = const <String, String>{}
+}) {
+  return interpolateStringList(
+      command,
+      Map<String, String>.of(additionalReplacementValues)
+        ..addAll(replacementValues)
+  );
+}
+
+
+class CustomDeviceLogReader extends DeviceLogReader {
+  CustomDeviceLogReader(this.name);
+
+  /// The name of the device this log reader is associated with.
+  @override
+  final String name;
+
+  final StreamController<String> _logLinesController = StreamController<String>.broadcast();
+
+  /// Listen to [process]' stdout and stderr, decode them using [SystemEncoding]
+  /// and add each decoded line to [logLines].
+  ///
+  /// However, [logLines] will not be done when the [process]' stdout and stderr
+  /// streams are done. So [logLines] will still be alive after the process has
+  /// finished.
+  ///
+  /// See [CustomDeviceLogReader.dispose] to end the [logLines] stream.
+  void listenToProcessOutput(Process process, {Encoding encoding = systemEncoding}) {
+    final Converter<List<int>, String> decoder = encoding.decoder;
+
+    process.stdout.transform<String>(decoder)
+      .transform<String>(const LineSplitter())
+      .listen(
```

Comment:
You can write this using a tear-off, here and elsewhere

`.listen(_logLinesController.add)`